go/enclave/rollupmanager/rollupmanager.go

```diff
@@ -0,0 +1,286 @@
+	b, err := re.storage.FetchHeadBlock()
+	if err != nil {
+		return nil, err
+	}
+	return re.getLatestRollupBeforeBlock(b)
+}
+
+func NextRollup(rollup *core.Rollup, batches []*core.Batch) *core.Rollup {
+	headBatch := batches[len(batches)-1]
+
+	rh := headBatch.Header.ToRollupHeader()
+
+	if rollup != nil {
+		rh.ParentHash = rollup.Header.Hash()
+	} else { // genesis
+		rh.ParentHash = gethcommon.Hash{}
+	}
+
+	rh.CrossChainMessages = make([]MessageBus.StructsCrossChainMessage, 0)
+	for _, b := range batches {
+		rh.CrossChainMessages = append(rh.CrossChainMessages, b.Header.CrossChainMessages...)
+	}
+
+	rollupHeight := big.NewInt(0)
+	if rollup != nil {
+		rollupHeight = rollup.Header.Number
+		rollupHeight.Add(rollupHeight, gethcommon.Big1)
+	}
+
+	rh.Number = rollupHeight
+	rh.HeadBatchHash = headBatch.Header.Hash()
+
+	return &core.Rollup{
+		Header:  rh,
+		Batches: batches,
+	}
+}
+
+func (re *RollupManager) CreateRollup() (*core.Rollup, error) {
+	rollup, err := re.fetchLatestRolluп()
+	if err != nil && !errors.Is(err, db.ErrNoRollups) {
+		return nil, err
+	}
+
+	hash := gethcommon.Hash{}
+	if rollup != nil {
+		hash = rollup.Header.HeadBatchHash
+	}
+
+	batches, err := re.l2chain.BatchesAfter(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(batches) == 0 {
+		return nil, fmt.Errorf("no batches for rollup")
+	}
+
+	if batches[len(batches)-1].Header.Hash() == hash {
+		return nil, fmt.Errorf("current head batch matches the rollup head bash")
+	}
+
+	newRollup := NextRollup(rollup, batches)
+
+	if err := re.l2chain.SignRollup(newRollup); err != nil {
+		return nil, err
+	}
+
+	return newRollup, nil
+}
+
+func (re *RollupManager) ProcessL1Block(b *types.Block) ([]*core.Rollup, error) {
+	return re.processRollups(b)
+}
+
+// ExtractRollups - returns a list of the rollups published in this block
+func (re *RollupManager) ExtractRollups(b *types.Block, blockResolver db.BlockResolver) []*core.Rollup {
+	rollups := make([]*core.Rollup, 0)
+	for _, tx := range b.Transactions() {
+		// go through all rollup transactions
+		t := re.MgmtContractLib.DecodeTx(tx)
+		if t == nil {
+			continue
+		}
+
+		if rolTx, ok := t.(*ethadapter.L1RollupTx); ok {
+			r, err := common.DecodeRollup(rolTx.Rollup)
+			if err != nil {
+				re.logger.Crit("could not decode rollup.", log.ErrKey, err)
+				return nil
+			}
+
+			// Ignore rollups created with proofs from different L1 blocks
+			// In case of L1 reorgs, rollups may end published on a fork
+			if blockResolver.IsBlockAncestor(b, r.Header.L1Proof) {
+				rollups = append(rollups, core.ToRollup(r, re.TransactionBlobCrypto))
+				re.logger.Trace(fmt.Sprintf("Extracted Rollup r_%d from block b_%d",
+					common.ShortHash(r.Hash()),
+					common.ShortHash(b.Hash()),
+				))
+			}
+		}
+	}
+
+	sort.Slice(rollups, func(i, j int) bool {
+		// Ascending order sort.
+		return rollups[i].Header.Number.Cmp(rollups[j].Header.Number) < 0
+	})
+
+	return rollups
+}
+
+// MOVED FROM L2 CHAIN
```

Comment:
Can probably get rid of this comment, won't make much sense to people after this PR :)